File: language/compiler/src/com/kubadziworski/parsing/visitor/expression/function/ArgumentExpressionVisitor.java
package com.kubadziworski.parsing.visitor.expression.function;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.ArgumentContext;
import jiva.parser.JivaParser.NamedArgumentContext;
import com.kubadziworski.domain.node.expression.Argument;
import com.kubadziworski.domain.node.expression.Expression;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

import java.util.Optional;

/**
 * Created by kuba on 09.05.16.
 */
public class ArgumentExpressionVisitor extends JivaBaseVisitor<Argument> {

    private final ExpressionVisitor expressionVisitor;

    public ArgumentExpressionVisitor(ExpressionVisitor expressionVisitor) {
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public Argument visitArgument(@NotNull ArgumentContext ctx) {
        Expression value = ctx.expression().accept(expressionVisitor);
        return new Argument(value, Optional.empty());
    }

    @Override
    public Argument visitNamedArgument(@NotNull NamedArgumentContext ctx) {
        Expression value = ctx.expression().accept(expressionVisitor);
        String name = ctx.name().getText();
        return new Argument(value, Optional.of(name));
    }

}

File: language/compiler/src/parsing/visitor/expression/function/ArgumentExpressionVisitor.kt
package jiva.parsing.visitor.expression.function

import jiva.domain.node.expression.Argument
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.ArgumentContext
import jiva.parser.JivaParser.NamedArgumentContext
import org.antlr.v4.runtime.misc.NotNull
import java.util.*

/**
 * Created by kuba on 09.05.16.
 */
class ArgumentExpressionVisitor(private val expressionVisitor: ExpressionVisitor) : JivaBaseVisitor<Argument>() {
    override fun visitArgument(@NotNull ctx: ArgumentContext): Argument {
        val value = ctx.expression().accept(expressionVisitor)!!
        return Argument(value, Optional.empty())
    }

    override fun visitNamedArgument(@NotNull ctx: NamedArgumentContext): Argument {
        val value = ctx.expression().accept(expressionVisitor)!!
        val name = ctx.name().text
        return Argument(value, Optional.of(name))
    }
}